### PR TITLE
:penguin: Add ubuntu arm rpi flavor

### LINKED ISSUE #645: :penguin: Add ARM/RPI support for Ubuntu 20.04 and 22.04 images
**Is your feature request related to a problem? Please describe.**
Currently kairos rpi images are available for alpine and opensuse. Would like ubuntu images to be added as well.

**Describe the solution you'd like**
ubuntu  20.04 and 22.04 rpi images for rpi 3 and 4

### PR DESCRIPTION
**Pull request by c0ffee**
**What this PR does / why we need it**:
Ubuntu flavors which support raspberrypi

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #645

### CODE REVIEW

**Comment by mudler** [comment]
Hey @c0ffee thanks! that's looking good at a first glance! I didn't had a deeper look on it yet, but will do as soon as possible - we need also to create the repository for the flavors as well before merging it

**Comment by c0ffee** [comment]
@mudler yeap the repos have to exist good point. so far it is booting on a rpi4, but would be good if someone with deeper knowledge of kairos can have a look :)

the main point is the size, because of the uncompressed firmwares the 2gb standard size for rpi img wont be enough, how should we handle this? simple increase the limit for all images?

**Comment by Itxaka** [comment]
> @mudler yeap the repos have to exist good point. so far it is booting on a rpi4, but would be good if someone with deeper knowledge of kairos can have a look :)
> 
> the main point is the size, because of the uncompressed firmwares the 2gb standard size for rpi img wont be enough, how should we handle this? simple increase the limit for all images?

Very nice!


Maybe we can compress those Fw files somehow? If that is the main difference maybe we can get away with it.

**Comment by c0ffee** [comment]
https://bugs.launchpad.net/ubuntu/+source/linux-firmware/+bug/1942260

not much movement there, maybe all are waiting for the zstd support in the kernel, but nevertheless this wouldnt help us on the lts releases :)

**Review by mudler — COMMENTED** [review]
It looks good here up to now, I'll give it another pass tomorrow and add the needed repositories :+1:

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-20-lts-arm-rpi:51]: There's no point to these lines as far as I can tell (and they are removed in #821).  `systemd-networkd` doesn't use them.

**Comment by mudler** [comment]
I've created the quay repository for the images (core and k3s):

- kairos-ubuntu-arm-rpi
- core-ubuntu-arm-rpi
- core-ubuntu-20-lts-arm-rpi
- kairos-ubuntu-20-lts-arm-rpi
- core-ubuntu-22-lts-arm-rpi
- kairos-ubuntu-22-lts-arm-rpi

File: images/Dockerfile.ubuntu-22-lts-arm-rpi
Comment (mudler) [inline images/Dockerfile.ubuntu-22-lts-arm-rpi:50]: ditto, same point from @sdwilsh

File: images/Dockerfile.ubuntu-arm-rpi
Comment (mudler) [inline images/Dockerfile.ubuntu-arm-rpi:48]: ditto

**Review by mudler — DISMISSED** [review]
This looks good overall here, thank you @c0ffee !

There are just some small nits that shouldn't block this PR from merging, as could be tackled separately as follow-ups.

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (c0ffee) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi:8]: @sdwilsh thanks for the sugesstions! i totally agree that we should clean this up, but i would prefer to do this as a second step and then for all ubuntu/debian variants

there are even a lot of packages that we should remove additionally, e.g. snapd? or why do we need nano and neovim

in general it would be good to define a list of packages that are needed to bring up a new kairos flavor and keep this to a minimum

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (c0ffee) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi:43]: did you verify that this is the only difference without "--no-install-recommends" ?

File: overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg
Comment (c0ffee) [reply overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg:6]: yeah i was lazy and copied it for now to get to a working image, in general it would be good to have a "group flavor" to avoid duplicates

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi:42]: There are a lot of packages that we grabbed, and I don't know of a way to get the full delta, if I'm being honest.  This just seemed like the minimum we needed to get the build working.

We can see what is listed as recommended in the build logs, and there's a bunch of stuff I don't think we actually need.  Then all those packages we don't install now also may have their own recommended additions, and that's harder to pull now.  In hindsight, I could look at logs from `master` and make a list of what we installed before and what we install now to get the full delta, but I'll have to do that tonight.

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-20-lts-arm-rpi:7]: This is a formatting nit-pick, but in the other ubuntu Docker files, the format is like this (with the `&&` after the newline):
```
RUN apt-get update \
  && ...
```

I don't know if there is a convention, but it'd be nice to be consistent.

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-20-lts-arm-rpi:18]: I just noticed this is true in the other debian/ubuntu ones as well, but is there a reason to have two packages on a single line?

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi:50]: Yeah, I'm just trying to make sure that it doesn't matter which order these PRs is landed :)

File: overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg
Comment (mudler) [reply overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg:6]: :+1: looks good here

**Review by mudler — DISMISSED** [review]
Looking good here :+1:

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (c0ffee) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi]: fixed

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (c0ffee) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi]: @sdwilsh you are fine with merging this for now like this and then do the needed changes after you are done with #821 ?

Comment (sdwilsh) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi]: Yeah, I can fix this up.  I'll spend some time on mine this weekend, but I'll have very minimal time soon as I have some friends visiting next week.  I'd rather see this get in before the next release than mine :D

Comment (c0ffee) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi]: yeah no problem only want to get this merged and then i can help you on the other one

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-20-lts-arm-rpi:61]: As you indicated before, we don't need to run `apt-get clean`

File: images/Dockerfile.ubuntu-22-lts-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-22-lts-arm-rpi:63]: no `apt-get clean` here either

File: images/Dockerfile.ubuntu-arm-rpi
Comment (sdwilsh) [inline images/Dockerfile.ubuntu-arm-rpi:61]: no `apt-get clean` here either

File: overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg
Comment (sdwilsh) [inline overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg:5]: old reference to "cOS".  This should be Kairos now, yeah?

File: overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml
Comment (sdwilsh) [inline overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml:22]: remove this line if it's commented out?

File: overlay/files-ubuntu-arm-rpi/system/oem/11_persistency.yaml
Comment (sdwilsh) [inline overlay/files-ubuntu-arm-rpi/system/oem/11_persistency.yaml:28]: We don't need to deal with this in this PR, IMO, but we will at some point: If a user wants to change something under `/boot/firmware/usercfg.txt`, that needs to be persisted across reboots.  With that said, we also probably want it to be read-only typically, so we need some way to configure it, and changes need to trigger a reboot.

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi:43]: I did move the `systemd-timesyncd` addition into #840, which has been merged, so you should add that to this and the other Ubuntu files as well.

**Review by sdwilsh — APPROVED** [review]

File: images/Dockerfile.ubuntu-20-lts-arm-rpi
Comment (sdwilsh) [reply images/Dockerfile.ubuntu-20-lts-arm-rpi]: I'm planning to try and get package parity with that one now, and then we can start cutting packages out, which is likely going to be much easier to reason about and verify that things are working correctly.

File: overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg
Comment (c0ffee) [reply overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg:5]: i copied simple the layer of the other flavor, i thought i better leave it all like it is

File: overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml
Comment (c0ffee) [reply overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml:22]: same like above, i hoped the original author left it there as a hint ;)

File: overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg
Comment (c0ffee) [reply overlay/files-ubuntu-arm-rpi/etc/cos/bootargs.cfg]: fixed

File: overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml
Comment (c0ffee) [reply overlay/files-ubuntu-arm-rpi/etc/elemental/config.yaml]: fixed

**Comment by sdwilsh** [comment]
Just a note: this is still a "draft" PR, although I think it's probably ready for review now :)

File: overlay/files-ubuntu-arm-rpi/system/oem/11_persistency.yaml
Comment (mudler) [reply overlay/files-ubuntu-arm-rpi/system/oem/11_persistency.yaml:28]: this can be customized by the user now directly from cloud config: #210  (not released yet, works only on master builds)

**Comment by c0ffee** [comment]
> Just a note: this is still a "draft" PR, although I think it's probably ready for review now :)

Yes I have to build all 3 images again and try it again on hardware, this is why I didn't change the status yet :)

**Comment by mauromorales** [comment]
Tried to build these for testing, but I keep hitting this error:

```
           +arm-image *failed* | ERROR Earthfile line 398:4
          +arm-image *failed* |       The command
          +arm-image *failed* |           WITH DOCKER RUN --privileged /build-arm-image.sh --model $MODEL --directory "/build/image" /build/$IMAGE_NAME
          +arm-image *failed* |       did not complete successfully. Exit code 11
```

https://github.com/mauromorales/kairos/actions/runs/4331798729/jobs/7563867150#step:8:6223

**Comment by sdwilsh** [comment]
@mauromorales, I think the relevant error from that log is a bit further up:
```
          +arm-image *failed* | >>> Copying files from /build/image
          +arm-image *failed* | rsync: [receiver] write failed on "/tmp/arm-builder.XQUJ45cub0/usr/src/linux-headers-5.15.0-1024-raspi/scripts/dtc/dtc-lexer.lex.c": No space left on device (28)
          +arm-image *failed* | rsync error: error in file IO (code 11) at receiver.c(378) [receiver=3.2.3]
          +arm-image *failed* | rsync: [sender] write error: Broken pipe (32)
```

**Comment by mauromorales** [comment]
Shoot, I missed that, thanks @sdwilsh!

**Comment by mauromorales** [comment]
Opened #1170 to test the new size changes contained in master. I'll close this one, but we can re-open anytime.

**Pull request closed without merging**
